### PR TITLE
Stricter locking assertions

### PR DESCRIPTION
**Pull request by nguyenhuy**
- Rename `ASDisplayNodeAssertLockUnownedByCurrentThread` to `ASAssertUnlocked`, and `ASDisplayNodeAssertLockOwnedByCurrentThread` to `ASAssertLocked` -> shorter and hopefully easier to distinguish between the two.
- Add assertions to `_locked_` and `_u_` (i.e "unlocked") methods.
- Turn `CHECK_LOCKING_SAFETY` on by default. After #1022 and #1023, we're in a good shape to actually enforce locked/unlocked requirements of internal methods. Our test suite passed, and we'll test more at Pinterest after the sync this week.
- Fix ASVideoNode to avoid calling `play` while holding the lock. That method inserts a subnode and must be called lock free.
- Other minor changes.

### CODE REVIEW

**Comment by nguyenhuy** [comment]
Note to reviewers: please review this **after** #1022 and #1023.

File: Source/ASDisplayNode+Layout.mm [hunk 1]
```diff
@@ -387,11 +394,14 @@ - (void)_u_measureNodeWithBoundsIfNecessary:(CGRect)bounds
     // In this case, we need to detect that we've already asked to be resized to match this
     // particular ASLayout object, and shouldn't loop asking again unless we have a different ASLayout.
     nextLayout->requestedLayoutFromAbove = YES;
-    __instanceLock__.unlock();
-    [self _u_setNeedsLayoutFromAbove];
-    __instanceLock__.lock();
+    
+    {
+      ASDN::MutexUnlocker u(__instanceLock__);
+      [self _u_setNeedsLayoutFromAbove];
```
Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: Apply the same fix in #1022 here just to be safe, i.e if `_u_setNeedsLayoutFromAbove` causes an assertion, it'll be correctly shown.

File: Source/ASDisplayNode+Layout.mm [hunk 2]
```diff
@@ -295,11 +297,12 @@ - (void)displayNodeDidInvalidateSizeNewSize:(CGSize)size
...
 - (void)_u_measureNodeWithBoundsIfNecessary:(CGRect)bounds
 {
-  ASDisplayNodeAssertLockUnownedByCurrentThread(__instanceLock);
+  ASDisplayNodeAssertLockNotHeld(__instanceLock__);
+  
   ASDN::MutexLocker l(__instanceLock__);
   // Check if we are a subnode in a layout transition.
   // In this case no measurement is needed as it's part of the layout transition
-  if ([self _isLayoutTransitionInvalid]) {
+  if ([self _locked_isLayoutTransitionInvalid]) {
```
Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: While having the lock, call the `_locked_` method.

File: Source/ASDisplayNode+Layout.mm [hunk 3]
```diff
@@ -530,18 +549,10 @@ - (void)transitionLayoutWithAnimation:(BOOL)animated
...
   ASDisplayNodeAssertMainThread();
-
-  ASSizeRange sizeRange;
-  {
-    ASDN::MutexLocker l(__instanceLock__);
-    sizeRange = [self _locked_constrainedSizeForLayoutPass];
-  }
-
-  [self transitionLayoutWithSizeRange:sizeRange
+  [self transitionLayoutWithSizeRange:[self _constrainedSizeForLayoutPass]
```
Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: Easier to just call the "unlocked" variant.

File: Source/ASDisplayNode.mm [hunk 4]
```diff
@@ -584,14 +590,20 @@ - (BOOL)isNodeLoaded
   if (ASDisplayNodeThreadIsMain()) {
     // Because the view and layer can only be created and destroyed on Main, that is also the only thread
     // where the state of this property can change. As an optimization, we can avoid locking.
-    return [self _locked_isNodeLoaded];
+    return [self _isNodeLoaded];
```
Comment (nguyenhuy) [inline Source/ASDisplayNode.mm]: Can't call the `_locked_` method here

File: Source/ASVideoNode.mm [hunk 5]
```diff
@@ -360,6 +361,8 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
         if (self.playerState == ASVideoNodePlayerStateLoading && _delegateFlags.delegateVideoNodeDidRecoverFromStall) {
           [self.delegate videoNodeDidRecoverFromStall:self];
         }
+        
+        ASUnlockScope(self);
```
Comment (nguyenhuy) [inline Source/ASVideoNode.mm]: As mentions in the description, `play` shouldn't be called with the lock.

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/9ac8581faf26687001a0aad4603377af/raw/99da0bbc13694a2061c0dc23567bc4cd4328c8e3/log.txt)

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/a19657f0f3b28abda6c9c7995933d9e4/raw/94e5ac58888019f5cbac31b2fe8a45afbd7cb5a9/log.txt)

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/a91bd32f0ee95540598be5f18a294b3b/raw/a068ef8e337a826be8b378a4bb7c113722fc60be/log.txt)

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/ab6b4305c63d870f1fc61ace1d8852ad/raw/e57baee6017bba6afebb79a8113a9701f0d11241/log.txt)

**Comment by nguyenhuy** [comment]
Uhh, I like that. Will update!

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/7d382ac5cabab85750ca2a23bd2e7b57/raw/90b17f2f0058ec3df49cd4d90a09b5164964c6e7/log.txt)

File: Source/ASDisplayNode+Layout.mm [hunk 6]
```diff
@@ -390,49 +396,60 @@ - (void)_u_measureNodeWithBoundsIfNecessary:(CGRect)bounds
...
 {
   // TODO: The logic in -_u_setNeedsLayoutFromAbove seems correct and doesn't use this method.
   // logic seems correct.  For what case does -this method need to do the CGSizeEqual checks?
   // IF WE CAN REMOVE BOUNDS CHECKS HERE, THEN WE CAN ALSO REMOVE "REQUESTED FROM ABOVE" CHECK
-  
+
+  ASAssertLocked(__instanceLock__);
+
   CGSize boundsSizeForLayout = ASCeilSizeValues(self.threadSafeBounds.size);
+  std::shared_ptr<ASDisplayNodeLayout> pendingLayout = _pendingDisplayNodeLayout;
```
Comment (garrettmoon) [inline Source/ASDisplayNode+Layout.mm]: Do you mind adding a comment explaining the shared_ptr's here?

File: Source/ASDisplayNode.mm (hunk 4 above)
Comment (garrettmoon) [reply Source/ASDisplayNode.mm]: Good catch!

File: Source/ASDisplayNode.mm [hunk 7]
```diff
@@ -584,14 +590,20 @@ - (BOOL)isNodeLoaded
...
 
 - (BOOL)_locked_isNodeLoaded
+{
+  ASAssertLocked(__instanceLock__);
+  return (_view != nil || (_layer != nil && _flags.layerBacked));
+}
+
+- (BOOL)_isNodeLoaded
 {
   return (_view != nil || (_layer != nil && _flags.layerBacked));
```
Comment (garrettmoon) [inline Source/ASDisplayNode.mm]: Does this need to be updated with a lock?

File: Source/ASVideoNode.mm (hunk 5 above)
Comment (garrettmoon) [reply Source/ASVideoNode.mm]: Is it worth adding a comment in code?

File: Source/ASDisplayNode.mm [hunk 8]
```diff
@@ -584,14 +590,21 @@ - (BOOL)isNodeLoaded
...
 }
 
 - (BOOL)_locked_isNodeLoaded
+{
+  ASAssertLocked(__instanceLock__);
+  return (_view != nil || (_layer != nil && _flags.layerBacked));
+}
+
+/// Called without the lock. Client is responsible for locking safety.
+- (BOOL)_unsafe_unlocked_isNodeLoaded
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode.mm]: Can we simplify all of this by just using `_layer != nil` inline when we want a fast loaded check?

Comment (nguyenhuy) [reply Source/ASDisplayNode.mm]: great point!

File: Source/ASDisplayNode.mm [hunk 9]
```diff
@@ -600,13 +600,13 @@ - (BOOL)isNodeLoaded
...
+  return _ASIsNodeLoaded(self);
 }
 
-/// Called without the lock. Client is responsible for locking safety.
-- (BOOL)_unsafe_unlocked_isNodeLoaded
+/// Can be called without the node's lock. Client is responsible for thread safety.
+ASDISPLAYNODE_INLINE AS_WARN_UNUSED_RESULT BOOL _ASIsNodeLoaded(ASDisplayNode *node)
 {
-  return (_view != nil || (_layer != nil && _flags.layerBacked));
+  return (node->_view != nil || (node->_layer != nil && node->_flags.layerBacked));
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode.mm]: What if we get rid of this function and just put `_layer != nil` inside the places where we need it?

**Pull request merged**